src/components/birth-date-picker/birth-date-picker.tsx

```diff
@@ -0,0 +1,115 @@
+// TODO: use MobX.
+// TODO: filter input?
+// TODO: should we update the state when we receive a different value?
+// TODO: date validation belongs in a separate library.
+
+import React = require("react");
+
+export interface BirthDatePickerProps {
+    value?: Date;
+    minDate?: Date;
+    maxDate?: Date;
+    onChange?: (newValue?: Date) => void;
+}
+
+export interface BirthDatePickerState {
+    valid: boolean;
+    year: string;
+    month: string;
+    day: string;
+}
+
+function isDateValid(date: Date): boolean {
+    return !Number.isNaN(date.getTime());
+}
+
+export function dateFromYearMonthDay(y: string, m: string, d: string): Date | undefined {
+    const looksPlausible = /^\d\d\d\d-\d\d?-\d\d?$/.test(`${y}-${m}-${d}`);
+    if (!looksPlausible) {
+        return undefined;
+    }
+
+    const date = new Date(`${y}-${m}-${d}Z`);
+    const valid = (
+        date.getUTCFullYear() === Number(y) &&
+        date.getUTCMonth()    === Number(m) - 1 &&
+        date.getUTCDate()     === Number(d)
+    );
+
+    return valid ? date : undefined;
+}
+
+export class BirthDatePicker extends React.Component<BirthDatePickerProps, BirthDatePickerState> {
+    static defaultProps: BirthDatePickerProps;
```

Comment:
This is a well written traditional React component. But, using the powers of mobx, we can rewrite it in much simpler, more declarative/reactive way.
1. No need for `.state` and `setState()`, no need for `ref`s.
2. The input fields update (from onChange) their respective `@observable` class properties
3. Have an `@autorun` validation method (which works with the observable class properties) and run it once in the constructor. The validation method triggers `this.props.onChange` if the date is valid.
4. That's it. Profit!